core/solver/minres.cpp

```diff
@@ -0,0 +1,299 @@
+    exec->run(minres::make_initialize(
+        r.get(), z.get(), p.get(), p_prev.get(), q.get(), q_prev.get(), v.get(),
+        beta.get(), gamma.get(), delta.get(), cos_prev.get(), cos.get(),
+        sin_prev.get(), sin.get(), eta_next.get(), eta.get(), &stop_status));
+
+    int iter = -1;
+    /* Memory movement summary:
+     * 27n * values + matrix/preconditioner storage
+     * 1x SpMV:           2n * values + storage
+     * 1x Preconditioner: 2n * values + storage
+     * 2x dot             4n
+     * 1x axpy            3n
+     * 1x step 1 (axpys)  16n
+     */
+    while (true) {
+        ++iter;
+        this->template log<log::Logger::iteration_complete>(
+            this, iter, nullptr, dense_x, tau.get(), nullptr);
+        if (stop_criterion->update()
+                .num_iterations(iter)
+                .residual(nullptr)
+                .residual_norm(tau.get())
+                .solution(dense_x)
+                .check(RelativeStoppingId, true, &stop_status, &one_changed)) {
+            break;
+        }
+
+        // Lanzcos (partial) update:
+        // v = A * z - beta * q_prev
+        // alpha = <v, z>
+        // v = v - alpha * q
+        // z_tilde = M * v
+        // beta = <v, z_tilde>
+        this->get_system_matrix()->apply(one_op.get(), z.get(),
+                                         neg_one_op.get(), v.get());
+        v->compute_conj_dot(z.get(), alpha.get());
+        v->sub_scaled(alpha.get(), q.get());
+        this->get_preconditioner()->apply(v.get(), z_tilde.get());
+        v->compute_conj_dot(z_tilde.get(), beta.get());
+
+        // Updates scalars (row vectors)
+        // finish Lanzcos:
+        // beta = sqrt(beta)
+        //
+        // apply two previous givens rotation to new column:
+        // delta = sin_prev * gamma  // 0 if iter = 0, 1
+        // tmp_d = gamma
+        // tmp_a = alpha
+        // gamma = cos_prev * cos * tmp_d + sin * tmp_a  // 0 if iter = 0
+        // alpha = -conj(sin) * cos_prev * tmp_d + cos * tmp_a
+        //
+        // compute and apply new Givens rotation:
+        // sin_prev = sin
+        // cos_prev = cos
+        // cos, sin = givens_rot(alpha, beta)
+        // alpha = cos * alpha + sin * beta
+        //
+        // apply new Givens rotation to eta:
+        // eta = eta_next
+        // eta_next = -conj(sin) * eta
+        //
+        // update residual norm approximation:
+        // tau = abs(sin) * tau
+        exec->run(minres::make_step_1(alpha.get(), beta.get(), gamma.get(),
+                                      delta.get(), cos_prev.get(), cos.get(),
+                                      sin_prev.get(), sin.get(), eta.get(),
+                                      eta_next.get(), tau.get(), &stop_status));
+
+
+        // update vectors
+        // update search direction and solution:
+        // swap(p, p_prev)
+        // p = (z - gamma * p_prev - delta * p) / alpha
+        // x = x + cos * eta * p
```

Comment:
No, this is correct, since eta is initialized differently to get the approximate `||r||`. This is again mentioned in https://web.stanford.edu/group/SOL/dissertations/sou-cheng-choi-thesis.pdf